contracts/src/NativeTokenDestination.sol

```diff
@@ -0,0 +1,313 @@
+     */
+    uint256 public immutable initialReserveImbalance;
+
+    /**
+     * @notice Tokens will not be minted until this value is 0 meaning the source contact is collateralized.
+     */
+    uint256 public currentReserveImbalance;
+
+    /**
+     * @notice Total number of tokens minted by this contract through the native minter precompile.
+     */
+    uint256 public totalMinted;
+
+    /**
+     * @notice The balance of BURNED_TX_FEES_ADDRESS the last time burned fees were reported to the source chain.
+     */
+    uint256 public latestBurnAddressBalance;
+
+    /**
+     * @notice tokenMultiplier allows this contract to scale the number of tokens it sends/receives to/from
+     * the source chain.
+     *
+     * @dev This can be used to normalize the number of decimals places between the tokens on
+     * the two subnets. Is calculated as 10^d, where d is decimalsShift specified in the constructor.
+     */
+    uint256 public immutable tokenMultiplier;
+
+    /**
+     * @notice If multiplyOnReceive is true, the raw token amount value will be multiplied by `tokenMultiplier` when tokens
+     * are transferred from the source chain into this destination chain, and divided by `tokenMultiplier` when
+     * tokens are transferred from this destination chain back to the source chain. This is intended
+     * when the "decimals" value on the source chain is less than the native EVM denomination of 18.
+     * If multiplyOnReceive is false, the raw token amount value will be divided by `tokenMultiplier` when tokens
+     * are transferred from the source chain into this destination chain, and multiplied by `tokenMultiplier` when
+     * tokens are transferred from this destination chain back to the source chain.
+     */
+    bool public immutable multiplyOnReceive;
+
+    /**
+     * @notice The wrapped native token contract that represents the native tokens on this chain.
+     */
+    IWrappedNativeToken public immutable token;
+
+    constructor(
+        address teleporterRegistryAddress,
+        address teleporterManager,
+        bytes32 sourceBlockchainID_,
+        address tokenSourceAddress_,
+        address feeTokenAddress_,
+        uint256 initialReserveImbalance_,
+        uint256 decimalsShift,
+        bool multiplyOnReceive_
+    )
+        TeleporterTokenDestination(
+            teleporterRegistryAddress,
+            teleporterManager,
+            sourceBlockchainID_,
+            tokenSourceAddress_,
+            feeTokenAddress_
+        )
+    {
+        token = IWrappedNativeToken(feeTokenAddress);
+
+        require(
+            initialReserveImbalance_ != 0, "NativeTokenDestination: zero initial reserve imbalance"
+        );
+
+        initialReserveImbalance = initialReserveImbalance_;
+        currentReserveImbalance = initialReserveImbalance_;
+
+        require(decimalsShift <= 18, "NativeTokenDestination: invalid decimalsShift");
+        tokenMultiplier = 10 ** decimalsShift;
```

Comment:
Is it possible to determine multiplyOnReceived based on decimalsShift, or do we have to provide this as a separate parameter?

Reply:
Oh, I suppose we could allow it to be specified as a signed integer. It does make the code a little messier though, I'm pretty ambivalent on this one

Reply:
Actually, I think `multiplyOnReceive` is clearer. It tells us which direction and what operation. Just passing in a signed int would be ambiguous.